libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,747 @@
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "mqtt_state.h"
+
+#define MQTT_PACKET_ID_INVALID ( uint16_t ) 0
+
+/**
+ * @brief Set the appropriate state when the packet is one of PUBACK,
+ * PUBREC, PUBREL, or PUBCOMP.
+ * 
+ * @param[in] opType Reserve, Send, or Receive.
+ * @param[in] idStatus Invalid, Unknown, or Known.
+ * @param[in] sendState The state to set for a SEND operation.
+ * @param[in] recvState The state to set for a RECEIVE operation.
+ * 
+ * @return The calculated state, which is one of MQTTInvalidState, `sendState`,
+ * or `recvState`.
+ */
+static MQTTPublishState_t _calculateAckState( MQTTStateOperation_t opType,
+                                              MQTTPacketIDState_t idStatus,
+                                              MQTTPublishState_t sendState,
+                                              MQTTPublishState_t recvState );
+
+/**
+ * @brief Calculate a state when QoS is 1.
+ *
+ * @param[in] packetType PUBLISH or PUBACK.
+ * @param[in] opType Reserve, Send, or Receive.
+ * @param[in] idStatus Invalid, Unknown, or Known.
+ *
+ * @return The calculated state.
+ */
+static MQTTPublishState_t _calculateQoS1( MQTTPublishType_t packetType,
+                                          MQTTStateOperation_t opType,
+                                          MQTTPacketIDState_t idStatus );
+
+/**
+ * @brief Calculate a state when QoS is 2.
+ *
+ * @param[in] packetType PUBLISH, PUBREC, PUBREL, or PUBCOMP.
+ * @param[in] opType Reserve, Send, or Receive.
+ * @param[in] idStatus Invalid, Unknown, or Known.
+ *
+ * @return The calculated state.
+ */
+static MQTTPublishState_t _calculateQoS2( MQTTPublishType_t packetType,
+                                          MQTTStateOperation_t opType,
+                                          MQTTPacketIDState_t idStatus );
+
+/**
+ * @brief Test if a transition to new state is possible.
+ *
+ * @param[in] currentState The current state.
+ * @param[in] newState State to transition to.
+ * @param[in] opType Reserve, Send, or Receive.
+ * @param[in] qos 0, 1, or 2.
+ *
+ * @note This function does not validate the current state, or the new state
+ * based on either the operation type or QoS. It assumes the new state is valid
+ * given the opType and QoS, which will be the case if calculated by
+ * _MQTT_CalculateState().
+ *
+ * @return `true` if transition is possible, else `false`
+ */
+static bool _validateTransition( MQTTPublishState_t currentState,
+                                 MQTTPublishState_t newState,
+                                 MQTTStateOperation_t opType,
+                                 MQTTQoS_t qos );
+
+/**
+ * @brief Test is a publish is outgoing or incoming.
```

Comment:
grammar: "Test __if__ a publish is outgoing or incoming."

Reply:
Fixed